### ISSUE TITLE
[GSoC] Improving `sympy.physics.control` : Phase 1

### ISSUE THREAD

**Issue by akshanshbhatt** [sympy/sympy#21539]
This phase will primarily involve completing the pending PR for the `TransferFunctionMatrix` class for MIMO systems. I will follow the [Documentation-Driven Development](https://gist.github.com/zsup/9434452) approach to complete this PR. This was suggested by @namannimmo10  in the meeting which we had yesterday. The steps involved will be-

1. Finalizing the characteristics of `TransferFunctionMatrix` and writing good examples for the users (covering all the cases). Being built upon a symbolic library, We can highlight the symbolic features in the examples.
2. Getting the examples verified from my mentors and also general feedback on the overall implementation.
3. Adding docstrings and writing unit tests before starting the development.
4. Coding according to the finalized docs. I will work in the continuation of PR #19761. It will preserve the commits made by Naman.

**Comment by moorepants** [comment]
That all sounds good to me. I back @namannimmo10 fully on the doc driven development. It is essential to develop some comprehensive examples first and then build the new class(es) to support the examples.

**Comment by namannimmo10** [comment]
Sounds good to me too! @akshanshbhatt, can you show an example through which we can get an idea about your proposed API for the transfer function matrix?

**Comment by akshanshbhatt** [comment]
@namannimmo10 I am working on it. I'll share it in a few hours.

**Comment by akshanshbhatt** [comment]
Hey @namannimmo10 and @moorepants, please look at this [gist](https://gist.github.com/akshanshbhatt/8d8175b6d899114da960c8dae934ed5a) and suggest improvements. It has the implementation details of the `TransferFunctionMatrix`'s API.

**Comment by akshanshbhatt** [comment]
@namannimmo10 @moorepants 
I am at present working on the textbook examples/problems. I will give an update on it very soon. Should I open a new pull request for the `TransferFunctionMatrix` now? 
As the WIP PR (#19761) has some minor merge conflicts, I plan to merge it to a new branch created from the current master and then open a PR. Working this way would preserve Naman's commits and resolve the issue with some of the tests failing on the CI (in #19761) due to obsolete dependencies that were removed in this period of 1 year.
If there is a better way, then please let me know.

**Comment by namannimmo10** [comment]
> Should I open a new pull request for the `TransferFunctionMatrix` now?

Yes! And it's better to take lots of feedback on the API design early on, so that the implementation can be done smoothly.

**Comment by akshanshbhatt** [comment]
@namannimmo10 @moorepants 
As planned, I have completed the textbook example on `TransferFunctionMatrix`. I have tried to include all the features discussed in the `TransferFunctionMatrix` API planning gist shared earlier. I have tried to make it as comprehensive as possible. It will be very much useful to general users who are new to this module. You can see the example [here](https://gist.github.com/akshanshbhatt/662dad40a797fdb78d44459354f3c79a) and review it.

**Comment by akshanshbhatt** [comment]
ping @namannimmo10
I am not accustomed to *rst* markup used for docs. The order in which the classes are ordered in https://docs.sympy.org/latest/modules/physics/control/lti.html is not correct, in my opinion. The order is not suitable for a beginner visiting this page for reference. `TransferFunction` should be on the top of the page for obvious reasons, instead of `Feedback`. How can I change this order?
In case you know how to fix this, please open a PR for the same.

**Comment by namannimmo10** [comment]
> I am not accustomed to rst markup used for docs

Check out the [sympy documentation style guide](https://github.com/sympy/sympy/blob/master/doc/src/documentation-style-guide.rst) (here you can also find a way to generate the documentation locally). SymPy uses sphinx for generating documentation, [this page](https://www.sphinx-doc.org/en/master/usage/extensions/autodoc.html) can give more useful info.

> How can I change this order?

I'm not sure why it doesn't respect the ordering of classes in `__all__`. But a quick fix I can think of right now is to update _lti.rst_ to:

```
===========
Control API
===========

lti
===

.. module:: sympy.physics.control.lti

.. autoclass:: TransferFunction
   :members:

.. autoclass:: Series
   :members:

.. autoclass:: Parallel
   :members:

.. autoclass:: Feedback
   :members:

```
(Or whatever ordering you'd like.)

**Comment by akshanshbhatt** [comment]
> I'm not sure why it doesn't respect the ordering of classes in `__all__`

I guess it was due to some default configuration in Sphinx's `automodule` to arrange the classes in alphabetical order unless specified otherwise.

Anyways, I have applied the changes suggested by you, and it seems to work fine now. I have opened a pull request (#21587) that you can review.

**Comment by akshanshbhatt** [comment]
@namannimmo10 
What is the purpose behind these unit tests for `TranferFunction`?

https://github.com/sympy/sympy/blob/eec3dd610b00970d522a90517d1251f87539167c/sympy/physics/control/tests/test_lti.py#L82-L92

In my opinion, float exponents shouldn't be allowed. It also contradicts the description given in the docstrings-

> A class for representing LTI (Linear, time-invariant) systems that can be **strictly** described by the ratio of polynomials in the Laplace Transform complex variable.

Should I change these tests to assert an error?

**Comment by namannimmo10** [comment]
I was actually skeptical about adding tests for float exponents; exponents must be whole numbers then only it will be a "polynomial".
See https://github.com/sympy/sympy/pull/19390#discussion_r439476048 . @moorepants also asked for the following functionality:

```python
G = TransferFunction(num, den, var)
G**(x**2 + 5*x - 2)
```
where `x != var` and `x` is a symbol representing an integer. What's your opinion about it?

**Comment by akshanshbhatt** [comment]
> where `x != var` and `x` is a symbol representing an integer. What's your opinion about it?

In that case, maybe we can impose an additional condition in `__pow__()` of `TransferFunction` to check for integer before raising the power. 

```python
>>> from sympy import Symbol
>>> x = Symbol("x", integer=True)
>>> print((x**2 + 55*x**98 + x + 89).is_integer)
True
```

`is_integer` seems to work fine for this purpose. What do you say?

Also, the comment you just mentioned is *exactly* a year old. What a coincidence :)

**Comment by akshanshbhatt** [comment]
@namannimmo10 @moorepants 

I've been working on a new implementation of the `TrasferFunctionMatrix` class, one that isn't a subclass of `ImmutableDenseMatrix`. Here is my new implementation, which works quite well. Here are some of the advantages -

* We will not have to handle multiplication and addition of `TransferFunctionMatrix` objects. These operations will have the efficiency of equivalent `ImmutableMatrix`.
* Only `TransferFunction` or `TransferFunction`-reducible objects (SISO `Series` and `Parallel`) are allowed. The time of object creation is reduced along with the lines of code. The old implementation had multiple sequential loops checking various conditions, which was very much inefficient.
* We will have control over the methods and attributes of the `ImmutableDenseMatrix` to include in the `TransferFunctionMatrix` class (rather than inheriting all).
* For Latex and Pretty-printing `TransferFunctionMatrix` objects, we won't have to depend on creating an external `Matrix` object.
* `obj.func(*obj.args)` works flawlessly. Earlier it was failing and had a `@XFAIL` decorator, which I have now removed.

  ```
  _________________________________________________________ xpassed tests _________________________________________________________ 
  sympy\core\tests\test_args.py: test_sympy__physics__control__lti__TransferFunctionMatrix
  ```

```python
class TransferFunctionMatrix(Basic):

    def __new__(cls, arg): 

        expr_mat_arg = []
        var = arg[0][0].var
        for row in range(len(arg)):
            temp = []
            for col in range(len(arg[row])):
                if not isinstance(arg[row][col], (TransferFunction, Series, Parallel)):
                    raise TypeError("Incompatible type found as the element of TransferFunctionMatrix")

                if var != arg[row][col].var:
                    raise ValueError("Conflicting value(s) found for `var`. All TransferFunction instances in "
                            "TransferFunctionMatrix should use the same complex variable in Laplace domain.")

                if isinstance(arg[row][col], TransferFunction):
                    temp.append(arg[row][col]._to_expr())
                else:
                    temp.append(arg[row][col].doit()._to_expr())
            expr_mat_arg.append(temp)
                

        if isinstance(arg, (tuple, list, Tuple)):
            # Making nested Tuple (sympy.core.containers.Tuple) from nested list or nested python tuple
            arg = Tuple(*(Tuple(*r, sympify=False) for r in arg), sympify=False)
        
        obj = super(TransferFunctionMatrix, cls).__new__(cls, arg)
        obj._expr_mat = ImmutableMatrix(expr_mat_arg)
        obj._var = arg[0][0].var
        obj._num_outputs, obj._num_inputs = obj._expr_mat.shape
        return obj
```

I thought it would be better to get suggestions before implementing them. Also, I will be chunking up the main PR into 2-3 smaller PRs for easing the review process. This PR will only include `TransferFunctionMatrix` and not the other MIMO aspects like `Series`, `Parallel`, and `Feedback` for MIMO systems. All these features, along with proper docs and tests, will be introduced in subsequent PRs.

**Comment by namannimmo10** [comment]
@akshanshbhatt, this looks promising, but I can't tell based on just the constructor whether this will work for sure or not. As long as it addresses all the concerns in the TFM PR, and you don't rewrite things from scratch, you're free to move with this.
